go/store/nbs/store.go

```diff
@@ -1635,26 +1640,154 @@ func (nbs *NomsBlockStore) WriteTableFile(ctx context.Context, fileId string, nu
+		return nil
+	}
+	var checkErr error
+	// TODO: Bigger batching.
+	checkDeps := func(c chunks.Chunk) {
+		if checkErr != nil {
+			return
+		}
+		addrs := make(hash.HashSet)
+		getAddrs(c)(ctx, addrs, func(hash.Hash) bool { return false })
+		remaining, err := refCheck(toHasRecords(addrs))
+		if err != nil {
+			checkErr = err
+		}
+		if len(remaining) > 0 {
+			checkErr = fmt.Errorf("%w, missing: %v", errors.New("cannot add table files; referenced chunk not found in store."), remaining)
+		}
+	}
+	for _, source := range sources {
+		err := source.iterateAllChunks(ctx, func(c chunks.Chunk) {
+			if checkErr != nil {
+				return
+			}
+			checkDeps(c)
+		}, nbs.stats)
+		if err != nil {
+			checkErr = errors.Join(err, checkErr)
+		}
+		if checkErr != nil {
+			break
+		}
+	}
+	return checkErr
+}
+
+func (nbs *NomsBlockStore) addTableFilesToManifest(ctx context.Context, fileIdToNumChunks map[string]int, getAddrs chunks.GetAddrsCurry, genRefCheck func([]hasRecord) (hash.HashSet, error)) error {
 	fileIdHashToNumChunks := make(map[hash.Hash]uint32)
 	for fileId, numChunks := range fileIdToNumChunks {
 		fileIdHash, ok := hash.MaybeParse(fileId)
-
 		if !ok {
 			return errors.New("invalid base32 encoded hash: " + fileId)
 		}
+		if numChunks > 0 {
+			fileIdHashToNumChunks[fileIdHash] = uint32(numChunks)
+		}
+	}
 
-		fileIdHashToNumChunks[fileIdHash] = uint32(numChunks)
-		totalChunks += numChunks
+	if len(fileIdHashToNumChunks) == 0 {
+		return nil
 	}
 
-	if totalChunks == 0 {
+	for {
+		// Open chunk sources for each of the fileIdHashToNumChunks files.
+		// This is done with the lock held. We also inspect |gcGen| at this
+		// time so that we can ensure our view of the store is still
+		// accurate when we add these files to the store if all the sanity
+		// checks pass.
+		chunkSources, gcGen, err := nbs.openChunkSourcesForAddTableFiles(ctx, fileIdHashToNumChunks)
+		if err != nil {
+			return err
+		}
+		// If these sources get added to the store, they will get cloned.
+		// Either way, we want to close these instances when we are done.
+		defer chunkSources.close()
+
+		refCheck := getRefCheck(nbs, genRefCheck, chunkSources)
+		err = refCheckAllSources(ctx, nbs, getAddrs, refCheck, chunkSources)
+		if err != nil {
+			// There was an error checking all references.
+			return err
+		}
+
+		// At this point, the added files are consistent with our view of the store.
+		// We add them to the set of table files in the store.
+		_, gcGenMismatch, err := nbs.updateManifestAddFiles(ctx, fileIdHashToNumChunks, nil, &gcGen, chunkSources)
+		if err != nil {
+			return err
+		} else if gcGenMismatch {
+			// A gcGenMismatch means that the store has changed out from under
+			// us as we were running these checks. We want to retry.
+			continue
+		}
+
+		// At this point, the files have been added to the
+		// store and the store was in a consistent state when
+		// we added them.
+		//
+		// We currently do not take any chunk dependencies
```

Comment:
GC is irrelevant here because we're checking the self-consistency of whole file, which additively will be consistent w/ GC results?